### DISCUSSION TITLE
Question on VTI based routing : site - site IPSec

### DISCUSSION THREAD

**Discussion by piddalagi** [strongswan/strongswan#909] [Q&A]
Dear all,

Since I wanted source IP based tunnelling I went through the below page.
https://wiki.strongswan.org/projects/strongswan/wiki/RouteBasedVPN

Extract from above page.

_It's important to note that VTI tunnel devices are a local feature, no additional encapsulation (like with GRE, [see below](https://wiki.strongswan.org/projects/strongswan/wiki/RouteBasedVPN#GRE)) is added, **so the other end does not have to be aware that VTI devices are used in addition to regular IPsec policies._**

I have topology as below:
                                          
    h1                                  R2                                 h3
        \                    tun1    /      \                              /
         s1 -------------------R1              R3------------------------s2
        /                    tun2    \      /                               \
    h2                                  R4                                  h4
                               
My queries:
Do I need to create VTI both on R1 and R3 OR on only one end suffices?
Is it mandatory to configure IP address for VTI?
Also is it mandatory that we need to disable rp_filter for VTI interface?

Your inputs will help me lot.

My configs are as follows.
Note: I have VTI config only on R1 as I need to route to specific tunnel (tun1 or tun1) based on source IP to R3. 

With below config, ping from h1 to h3 is failing. I am seeing below NoRoute errors on tun1.
  1 root@r1:/# ip -s tunnel show tun1
  2 tun1: ip/ip remote 20.0.1.2 local 10.0.1.1 ttl inherit key 1
  3 RX: Packets    Bytes        Errors CsumErrs OutOfSeq Mcasts
  4     0          0            0      0        0        0
  5 TX: Packets    Bytes        Errors DeadLoop NoRoute  NoBufs
  6     0          0            2      0        2        0
  7 root@r1:/# 

**R1:**

  1 root@r1:/# cat /etc/ipsec.conf
  2 config setup
  3         charondebug="all"
  4         uniqueids=yes
  5         strictcrlpolicy=no

  6 conn r1-r2-r3
  7         type=tunnel
  8         auto=start
  9         keyexchange=ikev2
 10         authby=secret
 11         left=10.0.1.1
 12         leftsubnet=172.0.1.0/24
 13         right=20.0.1.2
 14         rightsubnet=172.0.2.0/24
 15         ike=aes256-sha1-modp1024!
 16         esp=aes256-sha1!
 17         aggressive=no
 18         keyingtries=%forever
 19         ikelifetime=28800s
 20         lifetime=3600s
 21         dpddelay=30s
 22         dpdtimeout=120s
 23         dpdaction=restart
 24         mark=1

 25 conn r1-r4-r3
 26         type=tunnel
 27         auto=start
 28         keyexchange=ikev2
 29         authby=secret
 30         left=30.0.1.1
 31         leftsubnet=172.0.1.0/24
 32         right=40.0.1.2
 33         rightsubnet=172.0.2.0/24
 34         ike=aes256-sha1-modp1024!
 35         esp=aes256-sha1!
 36         aggressive=no
 37         keyingtries=%forever
 38         ikelifetime=28800s
 39         lifetime=3600s
 40         dpddelay=30s
 41         dpdtimeout=120s
 42         dpdaction=restart
 43         mark=2
 44 root@r1:/# 

  1 root@r1:/# cat /etc/strongswan.d/charon.conf | grep install_
  2     install_routes = no
  3     install_virtual_ip = no
  4     # install_virtual_ip_on =
  5 root@r1:/# 

VTI related commands:
ip tunnel add tun1 mode vti local 10.0.1.1 remote 20.0.1.2 key 1
ip link set up dev tun1
ip addr add 10.0.0.1/30 dev tun1
ip tunnel add tun2 mode vti local 30.0.1.1 remote 40.0.1.2 key 2
ip link set up dev tun2
ip addr add 10.0.0.2/30 dev tun2

**R3:**

 1 root@r3:/# cat /etc/ipsec.conf
  2 config setup
  3         charondebug="all"
  4         uniqueids=yes
  5         strictcrlpolicy=no

  6 conn r3-r2-r1
  7         type=tunnel
  8         auto=start
  9         keyexchange=ikev2
 10         authby=secret
 11         left=20.0.1.2
 12         leftsubnet=172.0.2.0/24
 13         right=10.0.1.1
 14         rightsubnet=172.0.1.0/24
 15         ike=aes256-sha1-modp1024!
 16         esp=aes256-sha1!
 17         aggressive=no
 18         keyingtries=%forever
 19         ikelifetime=28800s
 20         lifetime=3600s
 21         dpddelay=30s
 22         dpdtimeout=120s
 23         dpdaction=restart

 24 conn r3-r4-r1
 25         type=tunnel
 26         auto=start
 27         keyexchange=ikev2
 28         authby=secret
 29         left=40.0.1.2
 30         leftsubnet=172.0.2.0/24
 31         right=30.0.1.1
 32         rightsubnet=172.0.1.0/24
 33         ike=aes256-sha1-modp1024!
 34         esp=aes256-sha1!
 35         aggressive=no
 36         keyingtries=%forever
 37         ikelifetime=28800s
 38         lifetime=3600s
 39         dpddelay=30s
 40         dpdtimeout=120s
 41         dpdaction=restart
 42 root@r3:/#

**Comment by tobiasbrunner** [comment]
> Do I need to create VTI both on R1 and R3 OR on only one end suffices?

I guess that depends on your goals/requirements (e.g. if/how you want to control what SA the return traffic takes).

> Is it mandatory to configure IP address for VTI?

You have to specify addresses when creating it, but you don't need to add any addresses to it afterwards.

> Also is it mandatory that we need to disable rp_filter for VTI interface?

Mandatory, no. Whether it's necessary depends on the routes.

**Reply by piddalagi** [reply]
"what SA the return traffic takes" very precise and makes sense. For my requirement, I dont have any need of specific SA. So I am fine with VTI on left alone.

 Please consider my example:

_ip tunnel add tun1 mode vti local 10.0.1.1 remote 20.0.1.2 key 1
ip link set up dev tun1_

I have configured tun1 with peer and local with key matching the mark specified under ipsec.conf.

_ip tunnel add tun1 local 10.0.1.1 remote 20.0.1.2 mode vti key 1_

After this SA comes up fine with tunnel established. I have added route as below.

_ip route add 172.0.2.2/32 dev tun1_

Hope this means route the packet with dest IP 172.0.2.2 via tun1.
But with this in place, when I ping to 172.0.2.2 from lan-host (left subnet) I get "destination host unreachable" error.

_PING 172.0.2.2 (172.0.2.2) 56(84) bytes of data.
From 172.0.1.1 icmp_seq=1 Destination Host Unreachable
From 172.0.1.1 icmp_seq=2 Destination Host Unreachable
From 172.0.1.1 icmp_seq=3 Destination Host Unreachable_

Can you please point me where I have committed mistake in my config?

**Reply by tobiasbrunner** [reply]
Is `172.0.1.1` an IP address of R1? Is the `r1-r2-r3` connection fully established?

**Reply by piddalagi** [reply]
Hi Tobias,

You gave right hint. The r1-r2-r3 SA was not up. I did not notice, as I have two tunnels, some how I missed it.
The secret was missing on r3.

Very sorry for the trouble. Please mark this as answered.

Issue fixed!
